### PR TITLE
Move release script to a go file

### PR DESCRIPTION
**Pull request by landrito**
This moves the bulk of the release script to a go file. This also
changes the released binaries to become tarballs. This is for two
specific reasons:

1. By tarballing the executables, it preserves the +x permission on the
file.
1. It allows for simpler binary usage

### New
```
curl -L <link to tar> | sudo tar -zx -- --directory /usr/local/bin/
gapic-showcase start
```

### Old
```
curl -L <link to binary>
chmod u+x binary
mv binary /usr/local/bin/gapic-showcase
gapic-showcase start
```

### CODE REVIEW

File: util/release/main.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+func execute(args ...string) {
+	log.Print("Executing: ", strings.Join(args, " "))
+	if output, err := exec.Command(args[0], args[1:]...).CombinedOutput(); err != nil {
+		log.Fatalf("%s", output)
+	}
+}
+
+var sDirMu sync.Mutex = sync.Mutex{}
```
Comment (pongad) [inline util/release/main.go]: I think `sync.Once` is more appropriate here. Make the function update `showcaseDirMemo`, then 
```go
onceObj.Do(f)
return showcaseDirMemo
```

File: util/release/main.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	}
+	os.Chdir(showcaseDir())
+	execute("go", "get")
+	execute("go", "install")
+	version, err := exec.Command("gapic-showcase", "--version").Output()
+	if err != nil {
+		log.Fatal("Failed getting showcase version")
+	}
+
+	versionMemo = strings.TrimSpace(string(version[:]))
```
Comment (pongad) [inline util/release/main.go]: No need to `[:]`

File: util/release/main.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+func setup() {
+	os.Chdir(showcaseDir())
+
+	distDir := filepath.Join(showcaseDir(), "dist")
+	os.RemoveAll(distDir)
+	os.MkdirAll(distDir, 0755)
+
+	tmpDir := filepath.Join(showcaseDir(), "tmp")
+	os.RemoveAll(tmpDir)
+	os.MkdirAll(tmpDir, 0755)
```
[the remark this answers is not in the document]
Comment (landrito) [reply util/release/main.go]: Done!

File: util/release/main.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+	// Move showcase protos alongside it's dependencies.
+	protoDest := filepath.Join(
+		showcaseDir(),
+		"tmp",
+		"api-common-protos",
+		"google",
+		"showcase",
+		"v1alpha2")
+	os.MkdirAll(protoDest, 0755)
```
[the remark this answers is not in the document]
Comment (landrito) [reply util/release/main.go]: Done!

File: util/release/main.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+	tmpDir := filepath.Join(showcaseDir(), "tmp")
+	stagingDir := filepath.Join(tmpDir, fmt.Sprintf("gapic-showcase-v1alpha2-%s-protos", version()))
+	output := filepath.Join(showcaseDir(), "dist", fmt.Sprintf("gapic-showcase-v1alpha2-%s-protos.tar.gz", version()))
+
+	os.MkdirAll(stagingDir, 0755)
+	execute("cp", "-r", filepath.Join(tmpDir, "api-common-protos", "google"), stagingDir)
+
+	os.Chdir(tmpDir)
+	defer os.Chdir(showcaseDir())
+	execute("tar", "-zcf", output, fmt.Sprintf("gapic-showcase-v1alpha2-%s-protos", version()))
```
[the remark this answers is not in the document]
Comment (landrito) [reply util/release/main.go]: Ah I never knew about that option. That helps a lot!! Thanks.

File: util/release/main.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+	execute(append(command, files...)...)
+}
+
+func tarProtos() {
+	tmpDir := filepath.Join(showcaseDir(), "tmp")
+	stagingDir := filepath.Join(tmpDir, fmt.Sprintf("gapic-showcase-v1alpha2-%s-protos", version()))
+	output := filepath.Join(showcaseDir(), "dist", fmt.Sprintf("gapic-showcase-v1alpha2-%s-protos.tar.gz", version()))
+
+	os.MkdirAll(stagingDir, 0755)
+	execute("cp", "-r", filepath.Join(tmpDir, "api-common-protos", "google"), stagingDir)
```
[the remark this answers is not in the document]
Comment (landrito) [reply util/release/main.go]: The common protos directory contains some readme files that ideally I'd like to ignore. It's not much harm to keep them in the folder and its starting to seem that adding stuff to try and exclude the readmes is making this script messier. I'm just going to tar api-common-protos/google where they are like you suggested.

File: util/release/main.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+func compileBinaries() {
+	stagingDir := filepath.Join(showcaseDir(), "tmp", "x")
+	outputDir := filepath.Join(showcaseDir(), "dist")
+	execute("go", "get", "github.com/mitchellh/gox")
+
+	// This is a windows dependency that is not implicitly got since we only get
+	// the linux dependencies
+	execute("go", "get", "github.com/inconshreveable/mousetrap")
```
[the remark this answers is not in the document]
Comment (landrito) [reply util/release/main.go]: Done!

File: util/release/main.go [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+			"gox",
+			fmt.Sprintf("-osarch=%s", osArch),
+			"-output",
+			filepath.Join(stagingDir, fmt.Sprintf("gapic-showcase-v1alpha2-%s-{{.OS}}-{{.Arch}}/gapic-showcase", version())),
+			"github.com/googleapis/gapic-showcase")
+	}
+
+	dirs, _ := filepath.Glob(filepath.Join(stagingDir, "*"))
+	for _, dir := range dirs {
+		os.Chdir(dir)
```
[the remark this answers is not in the document]
Comment (landrito) [reply util/release/main.go]: Done!

File: util/release/main.go (hunk 1 above)
Comment (landrito) [reply util/release/main.go]: Done. Although, I did it using a closure. IDK if thats idiomatic in go.

File: util/release/main.go (hunk 2 above)
Comment (landrito) [reply util/release/main.go]: Done.

**Comment by landrito** [comment]
Thanks for all the feedback @pongad! This is ready for another look!

File: .circleci/config.yml [hunk 9]
```diff
@@ -90,33 +71,13 @@ jobs:
...
-                -o dist/gapic-showcase-v1alpha2-${VERSION}.desc \
-                tmp/api-common-protos/google/showcase/v1alpha2/echo.proto
-      - run:
-          name: Make a tarball of the protos.
+          name: Create release assets
           command: |
-            mkdir tmp/gapic-showcase-v1alpha2-${VERSION}-protos
-            cp -r tmp/api-common-protos/google tmp/gapic-showcase-v1alpha2-${VERSION}-protos
-            cd tmp/ && tar -zcf ../dist/gapic-showcase-v1alpha2-${VERSION}-protos.tar.gz gapic-showcase-v1alpha2-${VERSION}-protos
+            go get ./util/release
```
Comment (pongad) [inline .circleci/config.yml]: I don't think you actually need this line.

File: util/release/main.go [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+func executeInDir(dir string, args ...string) {
+	log.Printf("Executing in %s: %s", dir, strings.Join(args, " "))
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Dir = dir
+	if output, err := cmd.CombinedOutput(); err != nil {
+		log.Fatalf("%s", output)
+	}
+}
+
+var sOnce sync.Once = sync.Once{}
```
Comment (pongad) [inline util/release/main.go]: Just `var sOnce sync.Once` is good.

File: util/release/main.go [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+func setup() {
+	distDir := filepath.Join(showcaseDir(), "dist")
+	os.RemoveAll(distDir)
+	if err := os.MkdirAll(distDir, 0755); err != nil {
+		log.Fatalf("Failed to make the directory %s: %v", distDir, err)
+	}
+
+	tmpDir := filepath.Join(showcaseDir(), "tmp")
+	os.RemoveAll(tmpDir)
```
Comment (pongad) [inline util/release/main.go]: I think we should check errs here as well.

File: .circleci/config.yml (hunk 9 above)
Comment (landrito) [reply .circleci/config.yml]: Done.

File: util/release/main.go (hunk 10 above)
Comment (landrito) [reply util/release/main.go]: Done.

File: util/release/main.go (hunk 11 above)
Comment (landrito) [reply util/release/main.go]: Done.

**Comment by landrito** [comment]
Thanks for the review!

**Pull request merged**